### ISSUE TITLE
Roslyn fails to infer type arguments' nullable annotations from calls with lambda expressions even without closures

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#33211]
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 2.1
VisualStudio.16.Preview/16.0.0-pre.2.1+28529.54
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
class C
{
  public static void M()
  {
    var v = MyMethod(x =>
    {
      x = "";
      var result = GetT(x);
      result.ToString(); // no warnings
      // result = null; // CS8625 if uncommented
      return result;
    });

    v.ToString(); // incorrect CS8602
  }

  public static extern T MyMethod<T>(System.Func<string?, T> f);
}
``` 

**Expected Behavior**:
No warnings for the line `v.ToString();` since Roslyn is aware that the `result` variable is non-nullable. E.g. it warns about assigning possible null values to it and does not warn about dereferencing it.
Therefore the lambda expression with only non-nullable returns should be inferred to have a non-nullable return type and the `MyMethod.T` type argument for the call should be inferred to be a non-nullable string as well.

**Actual Behavior**:
Incorrect warning `CS8602 Possible dereference of a null reference.` is reported for the line `v.ToString();`

**Comment by gafter** [comment]
I am unable to repro this with a recent compiler build.
